File: backend/app/schemas/documents.py
"""文档与索引相关的请求体与响应体。

契约来源：`docs/DESIGN_IMPLEMENTATION.md` 第 6.3、6.4 节。
字段名与那里保持一致（如 `document_id` + `task_id`），前端类型再镜像一份。
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.document import DocumentStatus
from app.models.ingestion_task import IngestionStatus
from app.models.knowledge_base_index import IndexStatus


class DocumentResponse(BaseModel):
    """单个文档。"""

    model_config = ConfigDict(from_attributes=True)

    id: int
    filename: str
    content_type: str | None
    size_bytes: int
    status: DocumentStatus
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    """文档列表。遵守 `docs/CODING_CONVENTIONS.md` 第 6.3 节的列表响应规范。"""

    items: list[DocumentResponse]
    total: int
    page: int
    page_size: int
    # 各状态计数，前端用它渲染概览，避免再发一次请求
    status_summary: dict[str, int]


class UploadResponse(BaseModel):
    """上传响应。对应设计文档第 6.3 节的 `{document_id, task_id, status}`。"""

    document_id: int
    task_id: str
    status: str
    filename: str
    size_bytes: int


class TaskResponse(BaseModel):
    """摄取任务状态。对应设计文档第 6.3 节的 `{task_id, status, progress, error}`。"""

    model_config = ConfigDict(from_attributes=True)

    task_id: str
    document_id: int
    status: IngestionStatus
    progress: int
    error: str | None
    started_at: datetime | None
    finished_at: datetime | None


class RebuildResponse(BaseModel):
    """重建索引响应。对应设计文档第 6.4 节。"""

    # 用户没有任何文档时为 null，前端据此提示「请先上传文档」
    task_id: str | None
    status: str
    message: str
    document_count: int


class IndexStatusResponse(BaseModel):
    """索引状态。对应设计文档第 6.4 节。"""

    status: IndexStatus
    collection_name: str | None
    index_version: int | None
    embedding_model: str | None
    embedding_dimension: int | None
    chunk_size: int | None
    chunk_overlap: int | None
    document_count: int
    # 索引不可用时的可读原因，直接展示给用户
    message: str
    # 向量库与文档存储的连通性，便于用户排查「为什么索引建不起来」
    vector_store: dict[str, object]
    doc_store: dict[str, object]


class DeleteDocumentResponse(BaseModel):
    """删除文档响应。"""

    status: str = "success"
    message: str
    document_id: int


class PageQuery(BaseModel):
    """列表分页参数。参数名统一与 `docs/CODING_CONVENTIONS.md` 第 6.2 节一致。"""

    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=200)

File: frontend/src/types/documents.ts
/**
 * 文档与索引类型。
 *
 * 字段与 `docs/DESIGN_IMPLEMENTATION.md` 第 6.3、6.4 节的 API 契约一一对应，
 * 后端 `backend/app/schemas/documents.py` 是唯一来源。
 */

/** 文档状态。与后端 `DocumentStatus` 保持一致 */
export const DocumentStatus = {
  UPLOADED: "uploaded",
  INDEXED: "indexed",
  FAILED: "failed",
  DELETED: "deleted",
} as const;

export type DocumentStatusValue = (typeof DocumentStatus)[keyof typeof DocumentStatus];

/** 索引状态。与后端 `IndexStatus` 保持一致 */
export const IndexStatus = {
  BUILDING: "building",
  READY: "ready",
  STALE: "stale",
  FAILED: "failed",
} as const;

export type IndexStatusValue = (typeof IndexStatus)[keyof typeof IndexStatus];

/** 摄取任务状态。与后端 `IngestionStatus` 保持一致 */
export const IngestionStatus = {
  PENDING: "pending",
  RUNNING: "running",
  SUCCESS: "success",
  FAILED: "failed",
} as const;

export type IngestionStatusValue = (typeof IngestionStatus)[keyof typeof IngestionStatus];

/** 单个文档，对应 GET /api/docs 的 items 元素 */
export interface DocumentItem {
  id: number;
  filename: string;
  content_type: string | null;
  size_bytes: number;
  status: DocumentStatusValue;
  created_at: string;
  updated_at: string;
}

/** 文档列表响应 */
export interface DocumentListResponse {
  items: DocumentItem[];
  total: number;
  page: number;
  page_size: number;
  status_summary: Record<string, number>;
}

/** 上传响应，对应 POST /api/docs/upload */
export interface UploadResponse {
  document_id: number;
  task_id: string;
  status: IngestionStatusValue;
  filename: string;
  size_bytes: number;
}

/** 摄取任务状态，对应 GET /api/docs/tasks/{task_id} */
export interface TaskResponse {
  task_id: string;
  document_id: number;
  status: IngestionStatusValue;
  progress: number;
  error: string | null;
  started_at: string | null;
  finished_at: string | null;
}

/** 重建索引响应，对应 POST /api/index/rebuild */
export interface RebuildResponse {
  task_id: string | null;
  status: string;
  message: string;
  document_count: number;
}

/** 索引状态，对应 GET /api/index/status */
export interface IndexStatusInfo {
  status: IndexStatusValue;
  collection_name: string | null;
  index_version: number | null;
  embedding_model: string | null;
  embedding_dimension: number | null;
  chunk_size: number | null;
  chunk_overlap: number | null;
  document_count: number;
  message: string;
  vector_store: Record<string, unknown>;
  doc_store: Record<string, unknown>;
}

/** 删除文档响应 */
export interface DeleteDocumentResponse {
  status: string;
  message: string;
  document_id: number;
}

/** 前端本地维护的「进行中任务」，用于展示进度 */
export interface ActiveTask {
  taskId: string;
  documentId: number;
  filename: string;
  status: IngestionStatusValue;
  progress: number;
  error: string | null;
}
